models/perm/access/repo_permission.go

```diff
@@ -143,135 +143,162 @@ func (p *Permission) ColorFormat(s fmt.State) {
 	log.ColorFprintf(s, format, args...)
 }
 
-// GetUserRepoPermission returns the user permissions to the repository
-func GetUserRepoPermission(ctx context.Context, repo *repo_model.Repository, user *user_model.User) (perm Permission, err error) {
+// GetUserRepoPermissions returns the user permissions to the repositories
+func GetUserRepoPermissions(ctx context.Context, repos repo_model.RepositoryList, user *user_model.User) (perms map[int64]*Permission, err error) {
 	if log.IsTrace() {
 		defer func() {
 			if user == nil {
-				log.Trace("Permission Loaded for anonymous user in %-v:\nPermissions: %-+v",
-					repo,
-					perm)
+				log.Trace("Permissions Loaded for anonymous user in %-v:\nPermissions: %-+v",
+					repos,
+					perms)
 				return
 			}
-			log.Trace("Permission Loaded for %-v in %-v:\nPermissions: %-+v",
+			log.Trace("Permissions Loaded for %-v in %-v:\nPermissions: %-+v",
 				user,
-				repo,
-				perm)
+				repos,
+				perms)
 		}()
 	}
 
-	// anonymous user visit private repo.
-	// TODO: anonymous user visit public unit of private repo???
-	if user == nil && repo.IsPrivate {
-		perm.AccessMode = perm_model.AccessModeNone
-		return
+	perms = make(map[int64]*Permission, len(repos))
+	for _, repo := range repos {
+		perms[repo.ID] = &Permission{
+			AccessMode: perm_model.AccessModeNone,
+		}
 	}
 
-	var is bool
-	if user != nil {
-		is, err = repo_model.IsCollaborator(ctx, repo.ID, user.ID)
-		if err != nil {
-			return perm, err
+repoLoop:
+	for _, repo := range repos {
+		perms[repo.ID] = &Permission{
+			AccessMode: perm_model.AccessModeNone,
 		}
```

Comment:
Instead of always calling `perms[repo.ID]` we should perhaps extract a variable to call:
```suggestion
		repoPerm := &Permission{
			AccessMode: perm_model.AccessModeNone,
		}
		perms[repo.ID] = repoPerm
```
, and then refactor all subsequent calls to the new variable.
This is not only more readable, but should also be faster.